packages/graph/hash_graph/lib/graph/src/store/postgres/query/condition.rs

```diff
@@ -0,0 +1,328 @@
+        match self {
+            Condition::All(conditions) if conditions.is_empty() => fmt.write_str("true"),
+            Condition::Any(conditions) if conditions.is_empty() => fmt.write_str("false"),
+            Condition::All(conditions) | Condition::Any(conditions) => {
+                for (idx, condition) in conditions.iter().enumerate() {
+                    fmt.write_char('(')?;
+                    condition.transpile(fmt)?;
+                    fmt.write_char(')')?;
+                    if idx + 1 < conditions.len() {
+                        if matches!(self, Condition::All(_)) {
+                            fmt.write_str(" AND ")?;
+                        } else {
+                            fmt.write_str(" OR ")?;
+                        }
+                    }
+                }
+                Ok(())
+            }
+            Condition::Not(condition) => {
+                fmt.write_str("NOT(")?;
+                condition.transpile(fmt)?;
+                fmt.write_char(')')
+            }
+            Condition::Equal(value, None) | Condition::Equal(None, value) => {
+                value.transpile(fmt)?;
+                fmt.write_str(" IS NULL")
+            }
+            Condition::Equal(lhs, rhs) => {
+                lhs.transpile(fmt)?;
+                fmt.write_str(" = ")?;
+                rhs.transpile(fmt)
+            }
+            Condition::NotEqual(value, None) | Condition::NotEqual(None, value) => {
+                value.transpile(fmt)?;
+                fmt.write_str(" IS NOT NULL")
+            }
+            Condition::NotEqual(lhs, rhs) => {
+                lhs.transpile(fmt)?;
+                fmt.write_str(" != ")?;
+                rhs.transpile(fmt)
+            }
+        }
+    }
+}
+
+/// A [`FilterValue`] compiled to postgres.
+pub enum ConditionValue<'q> {
+    Column(Column<'q>),
+    Parameter(usize),
+}
+
+impl<'q> ConditionValue<'q> {
+    pub fn from_filter_value<'f: 'q, T: PostgresQueryRecord<'q>>(
+        value: &'f FilterValue<'q, T>,
+        parameters: &mut Vec<&'f dyn ToSql>,
+        alias: Option<TableAlias>,
+    ) -> Self {
+        match value {
+            FilterValue::Path(path) => Self::Column(Column {
+                table: Table {
+                    name: path.terminating_table_name(),
+                    alias,
+                },
+                access: path.column_access(),
+            }),
+            FilterValue::Parameter(parameter) => match parameter {
+                Parameter::Number(number) => {
+                    parameters.push(number);
+                    Self::Parameter(parameters.len())
+                }
+                Parameter::Text(text) => {
+                    parameters.push(text);
+                    Self::Parameter(parameters.len())
+                }
+                Parameter::Boolean(bool) => {
+                    parameters.push(bool);
+                    Self::Parameter(parameters.len())
```

Comment:
I can see that we do index the parameters from the queries correctly (starting from `$1` and not `$0`), but wondering where this is done? Is it because we always push to `parameters` before calling `.len()` in these match arms?

It may make sense to be explicit about the invariant somewhere as the value can be 0 as a `usize`